core/utilities-numbers.lua

```diff
@@ -0,0 +1,189 @@
+    -- There are valid cases where "@numbers=xxxx" with default(0) and decimal(1) styles both work.
+    -- Typically, with num=1234
+    --   "@numbers=arab" in default(0) --> ١٢٣٤
+    --   "@numbers=arab", in decimal(1) --> ١٬٢٣٤
+    -- But in many cases, ICU may fallback to latin, e.g. take "roman" (or "grek")
+    --   "@numbers=roman" in default(0) --> 1234
+    --   "@numbers=roman" in default(1) --> MCCXXXIV
+    -- Be user friendly and attempt honoring the script.
+    ok, result = pcall(icu.format_number, num, "@numbers=" .. options.system, 1)
+  end
+  if not ok then
+    SU.warn("Number formatting failed: " .. tostring(result))
+  end
+  return tostring(ok and result or num)
+end
+
+setmetatable (formatNumber, {
+  __call = function (self, num, options, case)
+    -- Formats a number according to options, and optional case
+    -- Options:
+    -- - system: a numbering system string, e.g. "latn" (= "arabic"), "roman", "arab", etc.
+    --   With the addition of "alpha".
+    --   Casing is guessed from the system (e.g. roman, Roman, ROMAN) unless specified
+    -- - style: a format style string, i.e. "default", "decimal", "ordinal", "string")
+    --   E.g. in English and latin script:   1234        1,234     1,124th    one thousand...
+    --   Possibly extended by additional language-specific formatting rules.
+    -- Obviously, some combinations of system, style and case won't do anything worth.
+    if math.abs(num) > 9223372036854775807 then
+      SU.warn("Integers larger than 64 bits do not reproduce properly in all formats")
+    end
+    options = options or {}
+
+    -- BEGIN COMPATIBILITY SHIM
+    if type(options) ~= "table" then
+      -- It used to be a string aggregating both concepts.
+      SU.deprecated("Previous syntax of SU.formatNumber", "new syntax for SU.formatNumber", "0.14.6", "0.16.0", [[
+  Previous syntax was SU.formatNumber(num, format[, case]) with a format string
+  New syntax is SU.formatNumber(num, options[, case]) with an options table,
+  possibly containing:
+    - system: a numbering system string, e.g. "latn" (= "arabic"), "roman", "arab", etc.
+      With the addition of "alpha".
+      Casing is taken into account (e.g. roman, Roman, ROMAN) unless specified
+    - style: a format style string, i.e. "default", "decimal", "ordinal", "string")
+      E.g. in English and latin script:   1234        1,234     1,124th    one thousand...
+      Possibly extended by additional language-specific formatting rules.
+  Note that the new syntax doesn't handle casing on the format style, for separation of
+  concerns.
+]])
+      if not case then
+        if options:match("^%l") then
+          case = "lower"
+        elseif options:match("^.%l") then
+          case = "title"
+        else
+          case = "upper"
+        end
+      end
+      if options:lower() == "nth" then
+        SU.deprecated("Format 'nth' in SU.formatNumber", "'ordinal' in SU.formatNumber", "0.14.6", "0.16.0")
```

Comment:
I think my only quibble with this whole PR is the naming fiasco for ordinal/ordinal-string/nth/cardinal etc. As I understand it the current shim you have in here will work to shim old documents to the new format, but it will also *fail* when old documents get updated because of the re-use of "ordinal" to mean something different. One would have to update documents exactly at the version boundary this shim was removed, not before. I don't think that's ideal at all.

At the moment I'm inclined to keep the "nth"/"ordinal" style names. As I understand it ICU doesn't have an equivalent to our "nth" style at all right?

Reply:
I suppose one way to deal with this is introduce a document level parameter that specifies the *target* SILE version. That might be handy for reference, but it could also allow us to automate cases like this where it is ambiguous whether the old behavior through a shim or new behavior is intended.

Reply:
Oh yuck. Thanks for setting the record straight.

Turkish has equivalents of both English forms, "1st" and "first" are "1." and "birinci", the period is not punctuation it is explicitly the cardinal marker and vocalized as "-inci".

What do you think about the backwards compatibility issue with the shim? Should I add a marker for documents that have been converted to target a specific API level or revert the scheme rename? Or some third option I'm missing?

Reply:
Nevermind the trailing question. I overlooked the argument changing from a string to a table, so that will be enough to know whether a shim should be used or not without conflating new/old meaning of "ordinal".

Reply:
> > As I understand it ICU doesn't have an equivalent to our "nth" style at all right?
> 
> It does: "ordinal" is the ICU name (`UNUM_ORDINAL`) for SILE's idiosyncratic English-sounding "nth". Both Eo and En had the correct ordinal _numbers_, resp. "1234-a" and "1234'th"[1](#user-content-fn-1-eb534ee3e1bd9c19d8f53e0201eb8748) here.

Oh, right. In that case the one it doesn't have is our "ordinal-string" combo right? It has "spellout" and "ordinal" but not a spelled out ordinal. Or am I still missing something? It seems to have an input parser for spelled out ordinals but I can't find a combination of output options to achieve it.

Reply:
Correct, it only has our "string" as `UNUM_SPELLOUT` (1000 = "one thousand"). It doesn't have spelt-out ordinals (but perhaps these are complex in the general case, or the Consortium just considered they would do it anyway... Too many language-specifics to address lol).

Also, by the way, note that ICU doesn't know anything about genders, plurals etc. (so for instances, 1 as ordinal is "1st" in English and "1e" in French - the latter only works actually for masculine singular.) Well, ICU can't be perfect I guess, for many reasons.